### ISSUE TITLE
Question: using username rather than first name and last name

### ISSUE THREAD

**Issue by jryd** [GeneaLabs/laravel-mixpanel#13]
Hey Mike,

I'm pretty new to using Mixpanel and have only set it up at the most basic level in each of my controllers - I'm looking to improve this and make it much more meaningful and this package looks perfect.

The only issue I have is that we only use a `username` and don't have `name`, `first_name` or `last_name` recorded for our users - as our application doesn't require it.

I can see from your Configuration steps that we can set a property to grab the `username` but this is then split out in to a `first_name` and `last_name`.

Is there anyway to use this package without that happening and track the People without using a first and last name, and just use their `username` for this instead?

I'm aware I could just build all of this myself, but with you having already set this up to listen for all the events I would like to it would be so much better if there was an easy way to just use this package.

Is there possibly an easy hack I could make to your code to do this, then I would just have to be careful not to update and lose that hack until I had got a more permanent solution done.

**Comment by simabo** [comment]
Hey jryd,

I'm in vacation and can't access my testing env right now, but you could do some test of your own in Mixpanel.

In your User Model, just write a [mutator](https://laravel.com/docs/5.2/eloquent-mutators#accessors-and-mutators), like this (provided that username is the field that is declared in your user table):
```
	public function getNameAttribute($value)
	{
		return $this->username;
	}

```

At runtime, the package does this ([see line 31](https://github.com/GeneaLabs/laravel-mixpanel/blob/master/src/Listeners/LaravelMixpanelUserObserver.php)):
```
        if ($user->name) {
            $nameParts = explode(' ', $user->name);
            array_filter($nameParts);
            $lastName = array_pop($nameParts);
            $firstName = implode(' ', $nameParts);
        }
        $data = [
            '$first_name' => $firstName,
            '$last_name' => $lastName,
            '$name' => $user->name,
            '$email' => $user->email,
            '$created' => ($user->created_at
                ? $user->created_at->format('Y-m-d\Th:i:s')
                : null),
        ];
```

Basically taking $user->name, which is actually username thanks to your mutator and splits it into first and last name. They will be sent, but you'll just have to ignore this info in MixPanel. Your username *will* be sent through $data['$name'].

I hope I didn't misunderstood your question.

**Comment by mikebronner** [comment]
Hi @jryd, can you explain why you would want to use the username only, and not the first-name/last-name?

This package will use the `name` field, which would be the username, and put it in the name field in Mixpanel. If it can't split the name into two, it will put the username in the first name field as well and leave the last name blank.

The intent is to identify humans by who they are, and not get tied up in what they represent in your system, so that anyone in the business without knowledge of the system can quickly identify what's going on. `SoftKittyy123` isn't likely to be recognized as a person by another human looking at it.

Prior to Laravel 5 there was the username field, but (in my opinion) using usernames is a bad practice, and I don't really want to do a whole lot of fiddling with them here. Perhaps if you can explain your use-case, I can see if it makes sense. :)

**Comment by jryd** [comment]
Hi @mikebronner, the reason I only want to use the username is because we don't store first or last name.

We're running a Laravel package to support account signups for an online game, and first and last names aren't required for this purpose and so we just use their username instead.

Given that it will just use their username instead, this will be fine and fit for purpose.

I've just gone through now to add the package in to Laravel and followed all of your installation/configuration steps:

1) `composer require genealabs/laravel-mixpanel:0.4.14`
2) Add service provider `'GeneaLabs\LaravelMixpanel\Providers\LaravelMixpanelServiceProvider',`
3) Add `MIXPANEL_TOKEN=xxxxxxxxxxxxxxxxxxxxxx` to my `.env` file and set the token
4) Add the `getNameAttribute()` mutator to my `User` model
5) Updated my `VerifyCsrfToken` middleware - it looks like Laravel 5.1.* now supports `protected $except` - to include `protected $except = [ 'genealabs/laravel-mixpanel/*', ];`

After doing all of this, I'm not seeing my live events updating from my test instance. I've tried running `php artisan cache:clear`, `php artican config:clear`, `php artisan clear-compiled` and `composer dump-autoload` but haven't been able to get it working. I've double checked and triple checked and my token is set correctly.

Any idea how to get it going, I did have this problem initially when I first began using Mixpanel with their PHP package - but I cannot remember how I got it to work in the end.

I appreciate your help Mike and @simabo!

**Comment by mikebronner** [comment]
Did you add `'model' => <YourNameSpace>\User::class,` to your auth.php config file? Do you see any error entries in your log file?

**Comment by jryd** [comment]
I already had the `'model'` set in my config files.

Playing around with it I have managed to get it to work - I had to manually add 

```
'mixpanel' => [
        'token' => env('MIXPANEL_TOKEN', 'add your token to your env file'),
        'enable-default-tracking' => true,
    ]
```

to my `config/services.php`. Straight after I did that it worked a treat!

**Comment by mikebronner** [comment]
Ah, yes! I abstracted that out after the 0.4 version. I will add that to the documentation, thanks! :)

**Comment by mikebronner** [comment]
Closing this for now, please reopen if you run into trouble. :)
